### DISCUSSION TITLE
How to create SEQUENCE OF?

### DISCUSSION THREAD

**Discussion by anthok** [librasn/rasn#16] [Q&A] [answered]
I've been reading the docs and looking through the repo for examples of using SEQUENCE OF but I haven't found much. 

I was thinking it would be something like this but not sure. Any help would be greatly appreciated!

```
#[derive(AsnType, Debug, Clone, Decode)]
pub struct example {
    #[rasn(tag(0))]
    var1: rasn::types::VisibleString,

    #[rasn(tag(1))]
    var2: rasn::types::Integer,

    #[rasn(tag(2))]
    data: Vec<Data>, // [2] IMPLICIT SEQUENCE OF Data
}

#[derive(AsnType, Debug, Clone, Decode)]
#[rasn(choice)]
pub struct Data {
    #[rasn(tag(0))]
    boolean: bool,
}

```

**Comment by XAMPPRocky** [comment, marked as answer]
Thank you for your question! It looks correct to me without seeing more. `Vec` is the Rust equivalent for `SEQUENCE OF`, you can also use it through the [`SequenceOf`](https://docs.rs/rasn/0.2.0/rasn/types/type.SequenceOf.html) type alias.

**Reply by anthok** [reply]
Thanks! I don't see `rasn::types::SequenceOf` in rasn 0.3.0, in the example above it doesn't error out but it just produces an empty `Vec`

**Reply by XAMPPRocky** [reply]
Oh whoops, I linked to old docs, though I probably should have added that back. Without seeing what data you're giving it I don't know what you would expect. Can you share a small test case?

**Reply by anthok** [reply]
Hopefully this works,  `800473737373810104a20c830100830100830100830100`

ASN1 BER
```
example ::= SEQUENCE {
        var1 	[0] IMPLICIT VisibleString,
	var2 	[1] IMPLICIT INTEGER,
	data        [2] IMPLICIT SEQUENCE OF Data,
}

Data ::= CHOICE{
	boolean		[0] IMPLICIT BOOLEAN,
}
```

**Reply by XAMPPRocky** [reply]
@anthok Can you send it to me in HEX, and also what value you expect the data to be?

**Reply by anthok** [reply]
Found potentially a better example, its only ASN1 after reserved 2.

```
0000   10 00 00 84 80 00 00 00 61 7a 80 23 73 69 6d 70
0010   6c 65 49 4f 47 65 6e 65 72 69 63 49 4f 2f 4c 4c
0020   4e 30 24 47 4f 24 67 63 62 45 76 65 6e 74 73 81
0030   02 00 c8 82 1d 73 69 6d 70 6c 65 49 4f 47 65 6e
0040   65 72 69 63 49 4f 2f 4c 4c 4e 30 24 45 76 65 6e
0050   74 73 83 06 65 76 65 6e 74 73 84 08 5c e2 7b e3
0060   c8 a0 5e 3f 85 01 01 86 01 01 87 01 01 88 01 01
0070   89 01 00 8a 01 04 ab 0c 83 01 00 83 01 00 83 01
0080   00 83 01 00
```
![Capture](https://user-images.githubusercontent.com/3751124/126520625-59ee992a-b036-49d1-99f6-be8cc54bb933.PNG)

Full ASN1
https://github.com/mz-automation/libiec61850/blob/v1.5/src/goose/iec61850_goose.asn


The `allData` field is the issue

**Reply by XAMPPRocky** [reply]
Sorry, I can't reproduce this. I can encode and decode entries in `all_data` just fine. See this test case I just added https://github.com/XAMPPRocky/rasn/blob/main/tests/iec61850.rs#L88 Can you reproduce your error with this code?

You can see the hex output of the encoding here http://lapo.it/asn1js/#YTCAAIECAMiCAIMGZXZlbnRzhACFAQGGAQGHAf-IAQGJAQCKAQSrCoMB_4MB_4UCAMg

**Reply by anthok** [reply]
Well I'm not actually sure yet what I did differently but it looks like it's definitely user error because your new unit test works for me. Thanks so much for all the help and love the library!
